modal/volume.py

```diff
@@ -192,5 +196,95 @@ async def read_file(self, path: Union[str, bytes]) -> AsyncIterator[bytes]:
             async for data in blob_iter(response.data_blob_id, self._client.stub):
                 yield data
 
+    @live_method
+    async def add_local_file(
+        self, local_path: Union[Path, str], remote_path: Optional[Union[str, PurePosixPath, None]] = None
+    ):
+        """
+        Upload a local file to the modal.Volume.
+
+        **Example:**
+
+        ```python notest
+        vol = modal.Volume.lookup("my-modal-volume")
+        vol.add_local_file("my-file.txt", "/remote-path/")
+        ```
+        """
+        mount_file = await self._upload_file(local_path, remote_path)
+        request = api_pb2.VolumePutFilesRequest(volume_id=self.object_id, files=[mount_file])
+        await retry_transient_errors(self._client.stub.VolumePutFiles, request, base_delay=1)
+
+    @live_method
+    async def add_local_dir(
+        self, local_path: Union[Path, str], remote_path: Optional[Union[str, PurePosixPath, None]] = None
+    ):
+        """
+        Upload a local directory to the modal.Volume.
+
+        **Example:**
+
+        ```python notest
+        vol = modal.Volume.lookup("my-modal-volume")
+        vol.add_local_dir("my-local-dir", "/remote-path/")
+        ```
+        """
+        _local_path = Path(local_path)
+        if remote_path is None:
+            remote_path = PurePosixPath("/", _local_path.name).as_posix()
+        else:
+            remote_path = PurePosixPath(remote_path).as_posix()
+
+        assert _local_path.is_dir()
+
+        def gen_transfers():
+            for subpath in _local_path.rglob("*"):
+                if subpath.is_dir():
+                    continue
+                relpath_str = subpath.relative_to(_local_path).as_posix()
+                yield self._upload_file(subpath, PurePosixPath(remote_path, relpath_str))
+
+        files = await ConcurrencyPool(20).run_coros(gen_transfers(), return_exceptions=True)
```

Comment:
I think we want to check if there's any exceptions in the returned list and raise an exception if there are.

Reply:
We want to abort early on exceptions, so `return_exception=False` - except that `asyncio.gather()` seems a bit wonky, in that it doesn't cancel other the other coros on exception:

```
    async def f(x):
        await asyncio.sleep(x)
        print(x)
        if x == 2:
            raise RuntimeError("bad")

    def gen():
        for i in range(0,5):
            yield f(i)

    try:
        g = asyncio.gather(*gen())
        await g
    except Exception as e:
        g.cancel()
        print("got exception")
        await asyncio.sleep(5)
```
-->
```
1
2
got exception
3
4
```

:exploding_head: 

Upon further inspection, `ConcurrencyPool` also seems completely broken - it doesn't limit concurrency at all.

Reply:
> it doesn't limit concurrency at all.

😖 good to know. guess that's a yak to shave

Reply:
Fixed here https://github.com/modal-labs/modal-client/pull/907/commits/f4b10817b4fd8e0340340cffc0c5cd30fd888751